extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala

```diff
@@ -17,63 +17,215 @@
-    super.beforeAll()
-    testUtils = new KPLBasedKinesisTestUtils
+  def createKinesisSourceForTest(testUtils: KPLBasedKinesisTestUtils): KinesisSource = {
+    new KinesisSource(
+      sqlContext,
+      testUtils.regionName,
+      testUtils.endpointUrl,
+      Set(testUtils.streamName),
+      InitialPositionInStream.TRIM_HORIZON,
+      SerializableAWSCredentials(KinesisTestUtils.getAWSCredentials()))
+  }
+}
+
+class KinesisSourceSuite extends KinesisSourceTest with KinesisFunSuite {
+
+  import testImplicits._
+
+  testIfEnabled("basic receiving and failover") {
+    var streamBlocksInLastBatch: Seq[StreamBlockId] = Seq.empty
+
+    def assertStreamBlocks: Boolean = {
+      if (sqlContext.sparkContext.isLocal) {
+        // Only test this one in local mode so that we can assume there is only one BlockManager
+        val streamBlocks =
+          SparkEnv.get.blockManager.getMatchingBlockIds(_.isInstanceOf[StreamBlockId])
+        val cleaned = streamBlocks.intersect(streamBlocksInLastBatch).isEmpty
+        streamBlocksInLastBatch = streamBlocks.map(_.asInstanceOf[StreamBlockId])
+        cleaned
+      } else {
+        true
+      }
+    }
+
+    val testUtils = new KPLBasedKinesisTestUtils
     testUtils.createStream()
-    streamName = testUtils.streamName
+    try {
+      val kinesisSource = createKinesisSourceForTest(testUtils)
+      val mapped =
+        kinesisSource.toDS[Array[Byte]]().map((bytes: Array[Byte]) => new String(bytes).toInt + 1)
+      val testData = 1 to 10
+      testStream(mapped)(
+        AddKinesisData(testUtils, kinesisSource, testData),
+        CheckAnswer((1 to 10).map(_ + 1): _*),
+        Assert(assertStreamBlocks, "Old stream blocks should be cleaned"),
+        StopStream,
+        AddKinesisData(testUtils, kinesisSource, 11 to 20),
+        StartStream,
+        CheckAnswer((1 to 20).map(_ + 1): _*),
+        Assert(assertStreamBlocks, "Old stream blocks should be cleaned"),
+        AddKinesisData(testUtils, kinesisSource, 21 to 30),
+        CheckAnswer((1 to 30).map(_ + 1): _*),
+        Assert(assertStreamBlocks, "Old stream blocks should be cleaned")
+      )
+    } finally {
+      testUtils.deleteStream()
+    }
   }
 
-  override def afterAll(): Unit = {
-    if (testUtils != null) {
+  testIfEnabled("DataFrameReader") {
+    val testUtils = new KPLBasedKinesisTestUtils
+    testUtils.createStream()
+    try {
+      val df = sqlContext.read
+        .option("region", testUtils.regionName)
+        .option("endpoint", testUtils.endpointUrl)
+        .option("stream", testUtils.streamName)
+        .option("position", InitialPositionInStream.TRIM_HORIZON.name())
+        .kinesis().stream()
+
+      val sources = df.queryExecution.analyzed.collect {
+          case StreamingRelation(s: KinesisSource, _) => s
+        }
+      assert(sources.size === 1)
+
+      // stream
+      assertExceptionAndMessage[IllegalArgumentException](
+        "Option 'stream' must be specified. Examples: " +
```

Comment:
I dont like exception message testing to be so specific, but I dont have strong feelings. I usually check for the keywords being present .. like the stream name.